src/python/grapl_analyzerlib/grapl_analyzerlib/query_gen.py

```diff
@@ -0,0 +1,545 @@
+            }}
+        }}
+    """
+
+    return vars_alloc, query
+
+
+def gen_query(
+    q: "Queryable", query_name: str, first: Optional[int] = None,
+) -> Tuple[VarAllocator, str]:
+    if not first:
+        first = 1
+
+    vars_alloc = VarAllocator()
+
+    func = find_func(q, vars_alloc)
+    var_query, var_block = into_var_query(q, "q0", vars_alloc, func=func, cascade=True)
+
+    merged = type(q)()
+    zip_graph(q, merged)
+    __merged_filters, merged_query_block = into_query_block(
+        merged, VarAllocator(), should_filter=False, should_alias=False,
+    )
+
+    vars_list = into_vars_list(vars_alloc)
+
+    f_first = ""
+    if first:
+        f_first = f", first: {first}"
+    query = f"""
+        query {query_name}({vars_list}) {{
+            {var_query}
+
+            {query_name}(func: uid(q0) {f_first}) @cascade {{
+                {merged_query_block}
+            }}
+        }}
+    """
+
+    return vars_alloc, query
+
+
+def traverse_query_iter(
+    root_q: "Queryable", visited: Optional[Set["Queryable"]] = None
+) -> Iterator["Queryable"]:
+    if visited is None:
+        visited = set()
+
+    if root_q in visited:
+        return
+    yield root_q
+
+    for edge_name, neighbor_filters in root_q.neighbor_filters():
+        if not neighbor_filters:
+            continue
+
+        visited.add(root_q)
+
+        for neighbor_filter in neighbor_filters:
+            if isinstance(neighbor_filter, tuple) or isinstance(neighbor_filter, list):
+                for n_filter in neighbor_filter:
+                    for nested in traverse_query_iter(n_filter, visited):
+                        yield nested
+            else:
+                for nested in traverse_query_iter(neighbor_filter, visited):
+                    yield nested
+
+
+def traverse_query_neighbors_iter(
+    root_q: "Queryable", visited: Optional[Set["Queryable"]] = None
+) -> Iterator[Tuple["Queryable", str, "EdgeFilter"]]:
+    if visited is None:
+        visited = set()
+
+    if root_q in visited:
+        return
+
+    for edge_name, neighbor_filters in root_q.neighbor_filters():
+        if not neighbor_filters:
+            continue
+        if root_q not in visited:
+            yield (root_q, edge_name, neighbor_filters)
+
+        visited.add(root_q)
+
+        for neighbor_filter in neighbor_filters:
+            if isinstance(neighbor_filter, tuple) or isinstance(neighbor_filter, list):
+                for n_filter in neighbor_filter:
+                    for nested in traverse_query_neighbors_iter(n_filter, visited):
+                        yield nested
+            else:
+                for nested in traverse_query_neighbors_iter(neighbor_filter, visited):
+                    yield nested
+
+
+from grapl_analyzerlib.queryable import Queryable, EdgeFilter
+from grapl_analyzerlib.comparators import (
+    Cmp,
+    Eq,
+    IntEq,
+    Has,
+    StrCmp,
+    IntCmp,
+    extract_value,
+    dgraph_prop_type,
+)
+
+
+def main():
```

Comment:
Is this a test? If so, can it be encoded as a regression test instead of manual?